### ISSUE TITLE
Mistake in FLOPs calculations in the paper

### ISSUE THREAD

**Issue by isayoften** [facebookresearch/blt#61]
Local models forwards at each byte, but debedder in BPE forwards at each token which is 4.4 times smaller in FLOPS. Considering your architecture hyper-parameters it seems like you missed that. In small models it gives advantage in terms of FLOPs for BLT for ±17%.

My calculations with proofs:
1) With your formulas: https://www.desmos.com/calculator/ippsrya4kf?lang=ru
2) With another formulas: https://colab.research.google.com/drive/1C3As0sG-SL482KslP-DxylpAYut-9nyM?usp=sharing #

**Comment by sriniiyer** [comment]
Thanks for looking deeply into this. Could you specify which equation in the paper has the issue?

**Comment by artidoro** [comment]
I believe you are referring to the compute optimal scaling experiments section. You are right that the models are not always matching in terms of FLOPs for the same size global model. That's why the results are best visualized as FLOPs/BPB plots and what matters is the scaling trend here.

**However, the advantage is not to BLT, if anything it might be favoring BPE!** Here is why.

We used compute optimal settings estimated for BPE from prior work. These are pretty good estimates of the optimal ratio of data to parameter count to get the best performing model for a fixed compute budget on our dataset. The result trend is a very strong baseline to beat. 

For BLT, we did not do a full search of the optimal data to parameter ratio. We simply added a local model on top of the transformer of the corresponding BPE model and trained for the same amount of data as the BPE model. As you pointed out, this ends up incurring in additional FLOPs depending on the scale of the LM.

We demonstrate **that there exist** a ratio of parameters to data for BLT that matches the BPE compute optimal ratio scaling trends. In fact, there might even be a better ratio!

Hope this helps!

**Comment by isayoften** [comment]
The equations in the paper are correct. However, the problem arises when I calculate the FLOPs using the architectural hyperparameters provided in the paper.

<img width="1221" alt="Image" src="https://github.com/user-attachments/assets/58bf990e-03a9-4e39-902d-c1d29a35f3f4" />

Let's calculate flops with formulas from the paper:

<img width="1237" alt="Image" src="https://github.com/user-attachments/assets/76d33904-d2ae-4e3a-aa3b-a9c434f375f4" />

assume our tokenizer has ps=4.5 and we use threshold that gives the same ps for BLT patcher; seqlen = 8192 bytes; 2B model

BPE_flops = 2 * 26 * 2 * 2560 * 4 * 2560(Feed-forward) + (1 * 2+2) * 2 * 26 * 2560^2(QKVO) + 4 * 26 * 2560 * ((8192/4.5+1)/2)(Attention) + 2 * 2560 * 128000(De-Embedding) = 4.98728e9

BPE_flops_per_byte = 4.98728e9/4.5 =  1.10828e9

Now let's calculate BLT flops:

<img width="1193" alt="Image" src="https://github.com/user-attachments/assets/b44c7582-9efd-4538-add8-2a5d9f8880c4" />

BLT_flops_per_byte = (2 * 26 * 2 * 2560 * 4 * 2560 + (1 * 2+2) * 2 * 26 * 2560^2 + 4 * 26 * 2560 * ((8192/4.5+1)/2))/4.5(global transformer) + ( 2 * 1 * 2 * 1024 * 4 * 1024 + (1 * 2+2) * 2 * 1 * 1024^2 + 4 * 1 * 1024 * ((512+1)/2) )(local encoder without cross attn) + ( 2 * 9 * 2 * 1024 * 4 * 1024 + (1 * 2+2) * 2 * 9 * 1024^2 + 4 * 9 * 1024 * ((512+1)/2) + 2 * 1024 * 256)(local decoder without cross attn) + (4 * 1 * 1024 * ((4.5+1)/2) + (2 * (4.5/3) + 2) * 2 * 1 * 1024^2) * (3/4.5) (encoder cross attn) + (4 * 9 * 1024 * ((3+1)/2) + (2 * (3/4.5) + 2) * 2 * 9 * 1024^2)  (decoder cross attn) = 1.29531e9

so (1.29531e9 - 1.10828e9)/(1.10828e9) = 0.168757

(same calculations here https://www.desmos.com/calculator/yuli2bdiud?lang=ru)

**Comment by isayoften** [comment]
Okay, that sounds fair, although some sense of unfairness still remains (if, for example, the scaling laws for BPE do indeed apply equally to BLT, as you, by the way, mention in the article, then without these extra FLOPs, the loss for BLT could potentially be worse, which still matters).

<img width="1194" alt="Image" src="https://github.com/user-attachments/assets/90ebee5e-4349-4585-b5e9-602e5dadc23c" />

Thank you so much for the answers!
